### PR TITLE
[flutter_tools] add un-discoverable preview device

### PR DESCRIPTION
**Pull request by jonahwilliams**
Add some basic tool to support a "flutter preview" device is not discoverable and doesn't have artifacts yet.

Work towards https://github.com/flutter/flutter/issues/85329

### CODE REVIEW

File: packages/flutter_tools/lib/src/preview_device.dart [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+      status = _logger.startProgress('Compiling application for preview...');
+      await _bundleBuilderFactory().build(
+        buildInfo: debuggingOptions.buildInfo,
+        mainPath: mainPath,
+        platform: TargetPlatform.tester,
+        assetDirPath: getAssetBuildDirectory(),
+      );
+      copyDirectory(_fileSystem.directory(
+        getAssetBuildDirectory()),
+      _assetDirectory.childDirectory('data').childDirectory('flutter_assets'),
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/preview_device.dart:122]: I think the whitespace is wrong but hard to tell in GitHub
```suggestion
        _assetDirectory.childDirectory('data').childDirectory('flutter_assets'),
```

File: packages/flutter_tools/lib/src/preview_device.dart [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+import 'project.dart';
+import 'protocol_discovery.dart';
+
+typedef BundleBuilderFactory = BundleBuilder Function();
+
+BundleBuilder _defaultBundleBuilder() {
+  return BundleBuilder();
+}
+
+class PreviewDevice extends Device {
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/preview_device.dart:33]: Can you add a doc comment giving a general overview of what this is for?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/preview_device.dart:38]: done

File: packages/flutter_tools/lib/src/preview_device.dart [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+typedef BundleBuilderFactory = BundleBuilder Function();
+
+BundleBuilder _defaultBundleBuilder() {
+  return BundleBuilder();
+}
+
+/// A device type that runs a prebuilt desktop binary alongside a locally compiled kernel file.
+///
+/// This could be used to support debug local development without plugins on machines that
+/// have not completed the SDK setup. These feautres are not fully implemented and the
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/preview_device.dart:36]:
```suggestion
/// have not completed the SDK setup. These features are not fully implemented and the
```

**Review by jmagman — APPROVED** [review]
LGTM for prototyping

**Pull request merged**